src/rcc.rs

```diff
@@ -96,43 +104,61 @@ impl CFGR {
         let pllm_min = (pllsrcclk + 1_999_999) / 2_000_000;
         let pllm_max = pllsrcclk / 1_000_000;
 
+        let target_freq = if self.pll48clk {
+            48_000_000
+        } else {
+            sysclk * sysclk_div
+        };
+
         // Find the lowest pllm value that minimize the difference between
-        // requested sysclk and actual sysclk.
-        let pllm = (pllm_min..=pllm_max).min_by_key(|pllm| {
-            let vco_in = pllsrcclk / pllm;
-            let plln = sysclk * sysclk_div / vco_in;
-            sysclk - (vco_in * plln / sysclk_div)
-        }).unwrap();
+        // target frequency and the real vco_out frequency.
+        let pllm = (pllm_min..=pllm_max)
+            .min_by_key(|pllm| {
+                let vco_in = pllsrcclk / pllm;
+                let plln = target_freq / vco_in;
+                target_freq - vco_in * plln
+            })
+            .unwrap();
 
         let vco_in = pllsrcclk / pllm;
         assert!(vco_in >= 1_000_000 && vco_in <= 2_000_000);
 
         // Main scaler, must result in >= 100MHz (>= 192MHz for F401)
         // and <= 432MHz, min 50, max 432
-        let plln = sysclk * sysclk_div / vco_in;
-
+        let plln = if self.pll48clk {
+            // try the different valid pllq according to the valid
+            // main scaller values, and take the best
+            let pllq = (4..=9)
+                .min_by_key(|pllq| {
+                    let plln = 48_000_000 * pllq / vco_in;
+                    let pll48_diff = 48_000_000 - vco_in * plln / pllq;
+                    let sysclk_diff = (sysclk as i32 - (vco_in * plln / sysclk_div) as i32).abs();
+                    (pll48_diff, sysclk_diff)
+                })
+                .unwrap();
+            48_000_000 * pllq / vco_in
+        } else {
+            sysclk * sysclk_div / vco_in
+        };
         let pllp = (sysclk_div / 2) - 1;
 
-        // Calculate real system clock
-        let sysclk = vco_in * plln / sysclk_div;
-
-        if sysclk != pllsrcclk {
-            // use PLL as source
-            rcc.pllcfgr.write(|w| unsafe {
-                w.pllm()
-                    .bits(pllm as u8)
-                    .plln()
-                    .bits(plln as u16)
-                    .pllp()
-                    .bits(pllp as u8)
-                    .pllsrc()
-                    .bit(self.hse.is_some())
-            });
-
-            (true, sysclk)
+        let pllq = (vco_in * plln) / 48_000_001 + 1;
```

Comment:
Why 48_000_001?

Reply:
Else you have to check for the case with exact division

Reply:
(i.e. we want to round upward, and adding 1 to the target division allow to round upward without test with a +1)

Reply:
Counterexample: `96_000_001 // 48_000_001 + 1 == 2`, should be 3
I think this should be something like `(A + N - 1) / N`

Reply:
That's so frustrating to not have the possibility to make unit tests :-(

Reply:
fixed thanks!